src/_guides/libraries/objective-c-interop.md

```diff
@@ -0,0 +1,224 @@
+interact with Objective-C APIs using only `dart:ffi`. But this involves a lot
+of boilerplate, so you can use `package:ffigen` to automatically generate
+the Dart FFI bindings for a given Objective-C API.
+
+## Example: AVAudioPlayer
+
+This guide will walk you through [an example]({{page.example}}) that uses
+`package:ffigen` to generate bindings for [`AVAudioPlayer`](
+{{page.appledoc}}/avfaudio/avaudioplayer?language=objc).
+
+Generating bindings to wrap an Objective-C API is similar to wrapping a C API.
+You point `package:ffigen` at the header file that describes the API, and then
+load the library with `dart:ffi`.
+
+`package:ffigen` parses your Objective-C header files using LLVM, so you'll
+need to install that first. See the [ffigen readme](
+{{page.ffigen}}#installing-llvm) for details.
+
+### The config file
+
+The first step is to configure ffigen. A minimal config file for Objective-C
+looks like this:
+
+```yaml
+language: objc
+output: 'foo_bindings.dart'
+headers:
+  entry-points:
+    - 'foo.h'
+```
+
+Alternatively, you can put this yaml in your pubspec.yaml file, under
+`ffigen:`.
+
+For our example, the entry point is the internal `AVAudioPlayer.h` header, and
+the output is `avf_audio_bindings.dart`. We've also set the `name:` and the
+`description:`.
+
+```yaml
+name: AVFAudio
+description: Bindings for AVFAudio.
+language: objc
+output: 'avf_audio_bindings.dart'
+headers:
+  entry-points:
+    - '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/AVFAudio.framework/Headers/AVAudioPlayer.h'
+```
+
+Another imortant thing you'll see, if you look at the
+[example config]({{page.example}}/pubspec.yaml), is a lot of exclusions.
+By default, ffigen will generate bindings for everything it finds
+in the header, and everything that those bindings depend on in other
+headers. Most Objective-C libraries depend on Apple's internal
+libraries, which are very large. If bindings are generated without any
+filters, the resulting file can be millions of lines long. To solve this
+problem, the ffigen config has fields that allow you to filter out
+all the functions, structs, enums etc that you're not interested in.
+For our case, we're only interested in `AVAudioPlayer`, and we exclude
+everything else.
+
+```yaml
+functions:
+  exclude:
+    - '.*'
+structs:
+  exclude:
+    - '.*'
+unions:
+  exclude:
+    - '.*'
+globals:
+  exclude:
+    - '.*'
+macros:
+  exclude:
+    - '.*'
+enums:
+  exclude:
+    - '.*'
+unnamed-enums:
+  exclude:
+    - '.*'
+objc-interfaces:
+  include:
+    - 'AVAudioPlayer'
+```
+
+Since we've included `AVAudioPlayer`, ffigen will exclude all other interfaces.
+The `exclude:` lines are all excluding the regex `'.*'`, which matches anything.
+The result is that nothing is included except `AVAudioPlayer`, and the
+things that it depends on, such as `NSObject`, `NSString`, etc. So instead
+of several million lines of bindings, we end up with tens of thousands.
```

Comment:
```suggestion
Since `AVAudioPlayer` is included, ffigen will exclude all other interfaces.
The `exclude` entries are all excluding the regular expressions `'.*'`, 
which matches anything.
The result is that nothing is included except `AVAudioPlayer`, 
and the things that it depends on, such as `NSObject` and `NSString`,
So instead of several million lines of bindings, 
we end up with tens of thousands.
```